### ISSUE TITLE
[Feature Request] Refactoring spec organisation

### ISSUE THREAD

**Issue by vmoens** [pytorch/rl#1140]
## Motivation

Specs are currently organised in a way that does not allow much flexibility for the reward, action and done specifications.
Here I propose a new organisation for the specs that allow these dedicated objects to live in nested structures if necessary.
The main use case for now is MARL where nesting the reward at some level will allow users to directly understand if the reward is shared amongst agents or distributed individually (same for done and actions).

### Example:

Shared reward:
```python
>>> batch = 3
>>> time = 20
>>> n_agents = 4
>>> features = 5
>>> actions = 2
>>> td = TensorDict({
...     "agent_features": TensorDict({
...         "observation": torch.zeros(batch, time, n_agents, features),
...         "action": torch.zeros(batch, time, n_agents, actions),
...     }, [batch, time, n_agents]),
...     "reward": torch.zeros(batch, time),
... }, [batch, time])
TensorDict(
    fields={
        agent_features: TensorDict(
            fields={
                action: Tensor(shape=torch.Size([3, 20, 4, 2]), device=cpu, dtype=torch.float32, is_shared=False),
                observation: Tensor(shape=torch.Size([3, 20, 4, 5]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([3, 20, 4]),
            device=None,
            is_shared=False),
        reward: Tensor(shape=torch.Size([3, 20]), device=cpu, dtype=torch.float32, is_shared=False)},
    batch_size=torch.Size([3, 20]),
    device=None,
    is_shared=False)
```

Distributed reward:
```python
>>> td = TensorDict({
...     "agent_features": TensorDict({
...         "observation": torch.zeros(batch, time, n_agents, features),
...         "action": torch.zeros(batch, time, n_agents, actions),
...         "reward": torch.zeros(batch, time, n_agents),
...     }, [batch, time, n_agents]),
... }, [batch, time])
TensorDict(
    fields={
        agent_features: TensorDict(
            fields={
                action: Tensor(shape=torch.Size([3, 20, 4, 2]), device=cpu, dtype=torch.float32, is_shared=False),
                observation: Tensor(shape=torch.Size([3, 20, 4, 5]), device=cpu, dtype=torch.float32, is_shared=False),
                reward: Tensor(shape=torch.Size([3, 20, 4]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([3, 20, 4]),
            device=None,
            is_shared=False)},
    batch_size=torch.Size([3, 20]),
    device=None,
    is_shared=False)
```

In this configuration, the level of nesting of the reward (or action, or done state) is indicative of its purpose.
This allows us to consider that whatever the action shape, for instance, we know that the tensordict that contains it is indicative of what should be considered as a "batch" for exploration purposes for instance. Hence, multi-dimensional actions will be allowed without any problem.

## Solution

To make this possible, I propose to refactor the specs in the following manner:

- `input_spec` will be renamed `_input_spec`. It will contain these specs and no other:
  - an `action_spec` object which will *always* be a `CompositeSpec`. The key(s) will represent where the action key will be found. In general, this will be a simple `_input_spec["action_spec"] = CompositeSpec(action=action_tensor_spec, shape=env.shape)`. The env will have an `action_key` attribute that will tell users where the `action_spec` leaf (`action_tensor_spec` in the example above) can be found inside the `action_spec` composite spec. Similarly, `env.action_spec` will return the leaf (and not the composite parent) such that the UX is not negatively impacted. In other words, `env.action_spec = OneHotSpec(...)` will still work as before.
  - a `state_spec` that contains all the input information that is not action.  

- 'output_spec` will be renamed `_output_spec`. It will contain these specs and no other:
  - a `reward_spec`. Again a composite spec. In the MARL example above, we would have `_output_spec["reward_spec"] = CompositeSpec(reward=smth, shape=env.shape)` for the first code snippet, and `_output_spec["reward_spec"] = CompositeSpec(agent_features=CompositeSpec(reward=smth, shape=(*env.shape, n_agents)), shape=env.shape)` in the second. As for `action_spec`, the `reward_spec` env attribute will point to the leaf and not the root to facilitate the interaction with this spec. For example, `env.reward_spec.is_in(tensordict.get(env.reward_key))` will work perfectly, and in most cases (reward at the root) this will be equivalent to `env.reward_spec.is_in(tensordict.get("reward"))`.
  - Same refactoring for `done_spec`
  - `observation_spec` will contain all output that is not `reward` and `done`. There is some discussion about the name `observation` that I agree is suboptimal as many output that are not reward or done are not observations (eg info, step counts etc). I'm open to a better name and a deprecation of `observation_spec` in favour of something else later. Candidates are `state_spec` or `output_spec` (but it'll be confusing with `_output_spec` at the root).

### Examples

Here are a few examples of what this will look like
```python
# if a regular spec is passed, it is wrapped in a composite spec
env.reward_spec = some_spec
assert env.reward_spec == some_spec
assert env.reward_spec == env._output_spec["reward_spec"][env.reward_key]

# if a composite spec is passed, it is not wrapped
env.reward_spec = CompositeSpec(reward=some_spec)
assert env.reward_spec == some_spec
assert env.reward_spec == env._output_spec["reward_spec"][env.reward_key]

env.action_spec = some_other_spec
assert env.action_spec == some_other_spec
# not sure about 'action_key', there might be more than one aciton entry?
assert env.action_spec == env_input_spec["action_spec"][env.action_key]

td = env.fake_tensordict()
td[env.action_key] # gets the action
td["next", *env.reward_key] # gets the reward
td["next", *env.done_key] # gets the done

td["next", "reward"] # gets the reward (in most cases)
td["next", "done"] # gets the done (in most cases)

```

As we can see this is minimally dirsuptive for existing users, and it hides "input_spec" and "output_spec" which are more confusing than anything else.

Little add-on: we could introduce a lock on the specs to avoid unwanted modifications of _input_spec and _output_spec.

cc @matteobettini

**Comment by matteobettini** [comment]
This is very cool, I think overall a nice change.

- I like that reward, done and action keys are becoming parameteric (with simple default). This is good generally as it will allow future flexibility. It will be a little bit painfull to adapt the rest of the library to it but maybe we can find an easy way to do it
- I like that the input and output spec are becoming fully hidden and i am pro to locking the user from directly touching those.
- i like the forced separation between shared and non shared marl keys

Some questions.

1. `state_spec` and `observation_spec` are still a bit biased imo. Your are saying: every extra input you need is in `state_spec` (but it might not be a state) and evey extra output you need goes in `observation_spec` (but it might not be an obs). Plus state_spec is by default not needed but observation_spec is by default needed. It seems a bit ununiform to me.  Maybe `state_spec` should be called input_spec and observation_spec should be called output_spec? or observation_spec stays and output_spec is introduced and by default empty (like state_spec). 
2. So when a value estimator will get a nested tensordict of this type, it will read done, reward and so on. Will we integrate default utils in the library files to expand the one with the least dimensions? I am saying this because I would like marl to be by default compatible with any loss. So if we carry a reward with less dimensionalty that the done (or vice-versa), we will have to implement some automatic mechanisms for expansion, without the users having to do manual things and mdoifying the tensordict betewen colection and training.
3. `# not sure about 'action_key', there might be more than one aciton entry?` couldn't the action_key lead you to a composite spec? as I understand not. This might be a problem. In other words, this method will constrain to leaf specs for all the values, it might be worth considering the implications of this. (I think this might already be a problem in the current library tho)

NIT: MARL actions always have to have the agent dimension IMO. I would argue that without per agent actions it is not MARL
NIT2: In your second snippet there is an error in the reward which should be nested

**Comment by vmoens** [comment]
> state_spec and observation_spec are still a bit biased imo. Your are saying: every extra input you need is in state_spec (but it might not be a state) and evey extra output you need goes in observation_spec (but it might not be an obs). Plus state_spec is by default not needed but observation_spec is by default needed. It seems a bit ununiform to me. Maybe state_spec should be called input_spec and observation_spec should be called output_spec? or observation_spec stays and output_spec is introduced and by default empty (like state_spec).

i'm open to rename observation_spec in output_spec (but maybe let's find another name for _output_spec?). Having an empty output_spec is not a great idea IMO. The only advantage I see is that it is good from a naming perspective (as in: this is an observation and this isn't), but for advanced or basic users of the library, it will be unclear if anything goes in output_spec or observation_spec. For instance: are images in Atari states or observations? Probably observations. If I normalize them, the new normalized entry should stay in observation no? What if I have a state? Should I explicitly tell my wrapper "this is an obs, this is an output"? What if keys collide in `output` and `observation`? It goes with a lot of extra work, design questions and very little improvement in UX or ease of use of the library IMO.

> So when a value estimator will get a nested tensordict of this type, it will read done, reward and so on. Will we integrate default utils in the library files to expand the one with the least dimensions? I am saying this because I would like marl to be by default compatible with any loss. So if we carry a reward with less dimensionalty that the done (or vice-versa), we will have to implement some automatic mechanisms for expansion, without the users having to do manual things and mdoifying the tensordict betewen colection and training.

Now that the time dimension is "marked" it will be easy to remove the constraints we have, at least in the value classes (not in the functionals). Something along the line of 
```python
time_dim = find_time_dim(tensordict)
reward = tensordict.get(reward_key)
done = tensordict.get(done_key)
reward, done = expand_given_time_dim(reward, done, time_dim)

def expand_given_time_dim(*tensors, time_dim):
    if tensors0.ndim != tensors1.ndim:
        new_shape = broadcast_shape(tensor0, tensor1)
        ...        
```
It's very sketchy but you get the gist. 

> couldn't the action_key lead you to a composite spec? as I understand not

Still poundering about that.
We could perfectly do it:
```
env.action_spec = OneHot(...) 
env.action_spec # retrieves the only leaf
env.action_spec = CompositeSpec(action1=OneHot(...), action2=ContinuousSpec(...))
env.action_spec # retrieves the composite spec because there is not a single leaf
```

> NIT2: In your second snippet there is an error in the reward which should be nested

good catch

**Comment by matteobettini** [comment]
`env.action_spec # retrieves the composite spec because there is not a single leaf`

So you would do that it retrives the leaf if there is only one and the composite if not? 
Or will it have something to do with the fact that the leaf is named "action"?

I think we have to well justify our policy for these choices:

Case 1: 
```python
env.x_spec = CompositeSpec(x=some_spec)
env.x_spec # some_spec
```
Case 2: 
```python
env.x_spec = some_spec
env.x_spec # some_spec
```
Case 3: 
```python
env.x_spec = CompositeSpec(x1=some_spec, x2=some_otehr_spec)
env.x_spec # CompositeSpec(x1=some_spec, x2=some_otehr_spec)
```
Case 4: 
```python
env.x_spec = CompositeSpec(x=some_spec, x2=some_otehr_spec)
env.x_spec # ???
```
What is our policy for case 4?

Case 5: 
```python
env.x_spec = CompositeSpec(agent_features=CompositeSpec(x=some_spec, x2=some_otehr_spec))
env.x_spec # ???
```
What is our policy for case 5?

there are many other examples I can construct where I do not know what level of nesting to return.
In general, when do we stop and decide if to return a leaf or a composite?

I think that independently of this we should not assume that x is a key in x_spec

**Comment by vmoens** [comment]
Oh no not the name "action". Only the fact that the composite spec content is not unique (there is more than one leaf, hence I don't know which one to give you, thereby you receive both)

**Comment by matteobettini** [comment]
ok so x_spec will return something different only when there is only one leaf in the composite tree?

**Comment by matteobettini** [comment]
Wouldn't this prevent MARL where agents have more action keys?

**Comment by vmoens** [comment]
why so?

**Comment by matteobettini** [comment]
In normal MARL 
```
env.action_spec = CompositeSpec(agent_features=CompositeSpec(action=some_spec))
env.action_spec # some_spec
```
which abstracts agent_features away

If agents have multiple actions
```
env.action_spec = CompositeSpec(agent_features=CompositeSpec(action1=some_spec, action2=some_other_spec))
env.action_spec # CompositeSpec(agent_features=CompositeSpec(action1=some_spec, action2=some_other_spec))
```
which doesn't abstracts agent_features away

Wouldn't this cause problems?

**Comment by vmoens** [comment]
not sure I see why this would cause problems, it would need adaptation of the code base here and there but you'd still do
```
env.action_spec.rand() # gives you a tensordict with {("agent_feature", "action1"), ("agent_feature", "action2")} keys 
```
What else would you want to see with 2 actions that can't be stored in a single tensor?

**Comment by matteobettini** [comment]
Ok so meybe for me to understand things better I need to clarify these 2 things:

- Why is 
```
env.action_spec = CompositeSpec(agent_features=CompositeSpec(action=some_spec))
env.action_spec # some_spec
```
Needed? couldn't it give the composite?

- When thinking of how this change fixes the problem with e-greedy, we said that with this change e-greedy can get the batch from the tensordict containing the action. If the action is in td["agent_features"] it will get the batch_size of that. Now, how does e-greedy now what is the tensordict containing the action? 
It seems like it would have to traverse the tensordict using the action_key and stop at the first biforcation or as soon as there are no more tds to traverse

**Comment by vmoens** [comment]
> Needed? couldn't it give the composite?

I believe this is (1) bc-breaking and (2) a major hustle for the 90% of users that have only one action at the root level. It will seem like an awkward thing for most. Most people will think that `env.action_spec.rand()` will return a single tensor, not a tensordict with a single `"action"` key. It is the most intuitive behaviour IMO. 

> When thinking of how this change fixes the problem with e-greed

If you look at `transforms.py` we have a util that looks at a method, and if the input is a CompositeSpec, it dispatches the method at its leaves. Here's how it works:
```
@_apply_to_composite
def fun(self, spec):
    new_spec = spec.foo()
    return new_spec
```
Now if `spec` is composite spec, we'll go through that composite spec and re-build it using each `spec.foo()` at the leaves.

In our case, it will be similar:
```
@_apply_to_composite
def fun(self, action_spec, tensordict, action_key=("action",)):
    assert isinstance(action_key, tuple) # no str allowed
    if len(action_key)>1:
        tensordict = tensordict.get(action_key[:-1])
        action_key = action_key[-1:]
    batch_shape = tensordict.shape
    rand = torch.zeros(batch_shape).bernoulli_(self.p)
    explorative_action = action_spec.rand()
    action = tensordict.get(action_key)
    action = expand_as_right(rand, action) * action + (1-expand_as_right(rand, action)) * explorative_action
    tensordict.set(action_key, action)
```

here `@_apply_to_composite` is different than before: if the action_spec is not composite, it does nothing. If it is, it calls this function with each of the leaves.

Would this work?

**Comment by matteobettini** [comment]
Ok that makes sense ty
